### PR TITLE
Bug Fix: Mutliple filter components overwrite each others conditions when filtering the same grid

### PR DESCRIPTION
**Pull request by lcarrasco-xdev**
**Steps to reproduce:**

1. Add two filter components which should filter the same grid to the ui.
2. Add a custom condition to the first filter component.
3. Add a custom condition to the second filter component.
4. The second condition overwrites the first one.

**Expected behavior:** At step 4 the second condition should not overwrite the first one. The second condition should be combined with the first condition.

### CODE REVIEW

File: vaadin-simple-grid-filter/src/main/java/software/xdev/vaadin/FilterComponent.java [hunk 1]
```diff
@@ -1500,4 +1504,65 @@ public FilterComponent<T> withInitialFilter(
...
+	
+	@Override
+	protected void onDetach(final DetachEvent detachEvent)
+	{
+		super.onDetach(detachEvent);
+		// Unregister from the event bus
+		this.registration.remove();
+	}
+	
+	private static class FilterComponentResetGridFilterEvent extends ComponentEvent<FilterComponent>
```
Comment (AB-xdev) [inline vaadin-simple-grid-filter/src/main/java/software/xdev/vaadin/FilterComponent.java:1561]: Type is missing, see sonar for more details

File: vaadin-simple-grid-filter/src/main/java/software/xdev/vaadin/FilterComponent.java [hunk 2]
```diff
@@ -938,10 +931,21 @@ private void removeChipBadgeCondition(final ChipBadgeExtension<FilterCondition<T
 	{
 		this.chipBadges.remove(chip);
 		this.hlChipBadges.remove(chip);
+		// Have to set all filter when removing a condition again because we cannot remove just one filter from the
+		// grid
 		this.updateGridFilter();
 		this.removeQueryParameter(chip);
 	}
 	
+	private void updateGridFilter()
```
Comment (AB-xdev) [inline vaadin-simple-grid-filter/src/main/java/software/xdev/vaadin/FilterComponent.java:940]: Method is called ``updateGridFilter`` but it just removes filters and throws a ``FilterComponentResetGridFilterEvent``.

Should likely be called ``resetGridFilters`` or?

File: vaadin-simple-grid-filter/src/main/java/software/xdev/vaadin/FilterComponent.java [hunk 3]
```diff
@@ -1500,4 +1504,65 @@ public FilterComponent<T> withInitialFilter(
...
+			final String inputValue = item1.getInputValue();
+			
+			predicates.add(
+				item1.getSelectedCondition().compare(
+					item1.getItem().getValueProvider(),
+					inputValue)
+			);
+		}
+		
+		return predicates.stream().map(p -> p.test(item)).reduce(Boolean::logicalAnd).orElseThrow();
```
Comment (AB-xdev) [inline vaadin-simple-grid-filter/src/main/java/software/xdev/vaadin/FilterComponent.java:1550]: Can likely be written with Streaming API

File: vaadin-simple-grid-filter/src/main/java/software/xdev/vaadin/FilterComponent.java [hunk 4]
```diff
@@ -1500,4 +1504,65 @@ public FilterComponent<T> withInitialFilter(
 		
 		return filterComponent;
 	}
+	
+	// Needed for interacting with other filter components on the same ui
+	private Registration registration;
```
Comment (AB-xdev) [inline vaadin-simple-grid-filter/src/main/java/software/xdev/vaadin/FilterComponent.java:1509]: Should be protected so that it can be overriden

**Review by AB-xdev — CHANGES_REQUESTED** [review]
Some general problems from my side:
1. I'm having some serious doubts if this change is needed in the first place. I think the use-case of needing multiple "filter components" for one Grid is extremely unlikely.
2. The current solution is not bound to the Grid that should be filter but instead to the UI, meaning that all "filter components" on the same UI are updated, which can result in unexpected behavior.
3. Self-registering an listener for communication between different components is not a good practice and breaks encapsulation. This will result in "works with magic"-behavior observed by users of the component.

**Comment by JohannesRabauer** [comment]
I think @AB-xdev has good points as to why we should not merge this change: https://github.com/xdev-software/vaadin-simple-grid-filter/pull/7#pullrequestreview-2381106280
That's why i'm closing this PR for now.

**Pull request closed without merging**